dali/operators/ssd/box_encoder.cc

```diff
@@ -80,74 +83,73 @@ vector<std::pair<unsigned, unsigned>> BoxEncoder<CPUBackend>::MatchBoxesWithAnch
   return matches;
 }
 
-vector<BoundingBox> BoxEncoder<CPUBackend>::ReadBoxesFromInput(
-  const float *in_boxes, unsigned num_boxes) const {
-  vector<BoundingBox> boxes;
-  boxes.reserve(num_boxes);
-  auto in_box_data = in_boxes;
-
-  for (unsigned idx = 0; idx < num_boxes; ++idx) {
-    boxes.push_back(BoundingBox::FromLtrb(in_box_data, BoundingBox::NoBounds()));
-    in_box_data += BoundingBox::kSize;
+template <int ndim>
+void WriteBoxToOutput(float *out_box_data, const vec<ndim, float> &center,
+                      const vec<ndim, float> &extent) {
+  for (int d = 0; d < ndim; d++) {
+    out_box_data[d] = center[d];
+    out_box_data[ndim + d] = extent[d];
   }
-
-  return boxes;
-}
-
-void BoxEncoder<CPUBackend>::WriteBoxToOutput(const std::array<float, BoundingBox::kSize> &box,
-                                              float *out_box_data) const {
-  for (unsigned idx = 0; idx < BoundingBox::kSize; ++idx)
-    out_box_data[idx] = box[idx];
 }
 
 void BoxEncoder<CPUBackend>::WriteAnchorsToOutput(float *out_boxes, int *out_labels) const {
   if (offset_) {
     std::memset(out_boxes, 0,
-      sizeof(std::remove_pointer<decltype(out_boxes)>::type)*BoundingBox::kSize*anchors_.size());
+                sizeof(std::remove_pointer<decltype(out_boxes)>::type) * BoundingBox::box_size *
+                    anchors_.size());
     std::memset(out_labels, 0,
-       sizeof(std::remove_pointer<decltype(out_labels)>::type)*anchors_.size());
+                sizeof(std::remove_pointer<decltype(out_labels)>::type) * anchors_.size());
   } else {
-    for (unsigned idx = 0; idx < anchors_.size(); ++idx) {
-      const auto box = anchors_[idx].AsCenterWh();
-      WriteBoxToOutput(box, out_boxes + idx * BoundingBox::kSize);
+    for (unsigned int idx = 0; idx < anchors_.size(); idx++) {
+      float *out_box = out_boxes + idx * BoundingBox::box_size;
+      const auto &anchor = anchors_[idx];
+      WriteBoxToOutput(out_box, anchor.centroid(), anchor.extent());
       out_labels[idx] = 0;
     }
   }
 }
 
 // Calculate offset from CenterWH ref box and anchor
 // based on eq (2)  in https://arxiv.org/abs/1512.02325 with extra normalization
-std::array<float, BoundingBox::kSize>
-GetOffsets(std::array<float, BoundingBox::kSize> box,
-           std::array<float, BoundingBox::kSize> anchor,
+std::pair<vec<2, float>, vec<2, float>>
+GetOffsets(vec<2, float> box_center,
+           vec<2, float> box_extent,
+           vec<2, float> anchor_center,
+           vec<2, float> anchor_extent,
```

Comment:
In `vec.h` we have `ivec2`, `dvec2` etc... I'd suggest adding `fvec` there. This would make the code way cleaner in this particular case

Reply:
done